pvlib/modelchain.py

```diff
@@ -810,32 +810,102 @@ def run_model(self, times=None, weather=None):
         times : None or DatetimeIndex, default None
-            Times at which to evaluate the model. Can be None if
-            attribute `times` is already set.
+            Times at which to evaluate the model. Can be None if attribute
+            `times` is already set.
         weather : None or DataFrame, default None
-            If None, assumes air temperature is 20 C, wind speed is 0
-            m/s and irradiation calculated from clear sky data. Column
-            names must be 'wind_speed', 'temp_air', 'dni', 'ghi', 'dhi'.
-            Do not pass incomplete irradiation data. Use method
-            :py:meth:`~pvlib.modelchain.ModelChain.complete_irradiance`
-            instead.
+            If None, assumes air temperature is 20 C, wind speed is 0 m/s and
+            irradiation calculated from clear sky data.
+            For conventional weather data, column names must be 'wind_speed',
+            'temp_air', 'dni', 'ghi', 'dhi'. Do not pass incomplete irradiation
+            data, instead use method
+            :py:meth:`~pvlib.modelchain.ModelChain.complete_irradiance`.
+            For complete PV reference device data, column names must be 'F' and
+            'H', for effective irradiance ratio and effective temperature
+            ratio, respectively. See :py:func:`~pvlib.pvsystem.sdm_campanelli`.
+            If either 'F' or 'H' is missing, then convential weather data are
+            also required to compute the missing 'F' or 'H'.
 
         Returns
         -------
         self
 
-        Assigns attributes: times, solar_position, airmass, irradiance,
-        total_irrad, effective_irradiance, weather, temps, aoi,
-        aoi_modifier, spectral_modifier, dc, ac, losses.
+        For conventional weather data, assigns attributes: times,
+        solar_position, airmass, irradiance, total_irrad, effective_irradiance,
+        weather, temps, aoi, aoi_modifier, spectral_modifier, dc, ac, losses.
+        For complete reference device data, only assigns attributes: times, dc,
+        ac, losses.
         """
 
-        self.prepare_inputs(times, weather)
-        self.aoi_model()
-        self.spectral_model()
-        self.effective_irradiance_model()
-        self.temp_model()
+        if weather is not None and 'F' in weather:
+            # Use PV reference device weather information for F
+            self.F = weather['F']
+        else:
+            # Clear any stale value
+            if hasattr(self, 'F'):
+                del self.F
+        if weather is not None and 'H' in weather:
+            # Use PV reference device weather information for H
+            self.H = weather['H']
+        else:
+            # Clear any stale value
+            if hasattr(self, 'H'):
+                del self.H
+        if not hasattr(self, 'F') and not hasattr(self, 'H') or \
+           hasattr(self, 'F') and not hasattr(self, 'H') or \
+           not hasattr(self, 'F') and hasattr(self, 'H'):
```

Comment:
Can anyone figure out a simplified equivalent logic?

Reply:
A whole set of functions do not need to be run in some cases, and I think they fail in the case where there is no data for them to run.

Reply:
Fair points, but I think `prepare_inputs` should be protection against missing values - you can even pass `weather=None` and it will fill in with a clear-sky model. I realize that `sdm_campanelli` doesn't need anything upstream of `dc_model` but I'd prefer to see the model chain sequence maintained.

My instinct is that an `if...then` structure to skip model chain steps for a choice of a downstream model will be a problem to maintain, and will become difficult to manage as more models are added.

I would prefer a separate pull request to extend the `ModelChain` functions to handle measured `effective_irradiance` and `temp_module` as inputs in a general manner.  I think this pull request should conclude when `sdm_campanelli` works as a method of the current `ModelChain`, with supporting functions consistent with the current patterns.  I'm OK with less efficient code in favor of consistency.  Redesign of `ModelChain` and generalizing treatment of input options can be opened as their own issues.

Reply:
Is this a reasonable compromise?

```python 
if self.dc_model == sdm_campanelli and 'F' in weather.columns and 'H' in weather.columns:
    mc.F = weather['F']
    mc.H = weather['H']
else:
    self.pepare_inputs ... self.temp_model()
self.dc_model()
...
```

Reply:
This makes sense to me. At present the only model that uses the `F` and `H` inputs is Campanelli, but it is possible that `F` is more generally useful in place of `effective_irradiance` calculated from weather in, e.g., SAPM.

Reply:
@cwhanse I'm pretty sure that `F` (and `H`) may be generalized to "all" the single diode models, the only difference being how they propagate into the "auxiliary equations". The derivation is analogous for double-diode models as well. One of my (longer term) goals for pv-fit.com is to fit calibration data to multiple models at once and provide the user with a model "ranking" based on a goodness of fit metric.

Would you be able to refer me to where in the literature the SAPM definition of "effective irradiance" is derived, esp. with respect to the conventional meteorological datasets?

Reply:
@thunderfish24 here: http://prod.sandia.gov/techlib/access-control.cgi/2004/043535.pdf

It's poorly explained. But, in a nutshell, `effective irradiance` is defined in the text as "the fraction of the total solar irradiance incident on the module to which the cells inside actually respond", i.e., define 'respond'. `Ee` is calculated from weather by Eq. 1 (to get Isc) then Eq. 7 (ratio of calculated Isc to a reference value). You can do the algebra and get an expression for `Ee` in terms of beam and diffuse on the array's surface. We often substitute measurements from an appropriate reference cell for the calculated `Ee` value.